internal/offline/check.go

```diff
@@ -0,0 +1,138 @@
+package offline
+
+import (
+	"fmt"
+	"os"
+	"path"
+
+	"github.com/google/osv-scanner/pkg/lockfile"
+	"github.com/google/osv-scanner/pkg/models"
+	"github.com/google/osv-scanner/pkg/osv"
+	"github.com/google/osv-scanner/pkg/reporter"
+)
+
+const zippedDBRemoteHost = "https://osv-vulnerabilities.storage.googleapis.com"
+const envKeyLocalDBCacheDirectory = "OSV_SCANNER_LOCAL_DB_CACHE_DIRECTORY"
+
+func loadDB(dbBasePath string, ecosystem lockfile.Ecosystem, offline bool) (*ZipDB, error) {
+	return NewZippedDB(dbBasePath, string(ecosystem), fmt.Sprintf("%s/%s/all.zip", zippedDBRemoteHost, ecosystem), offline)
+}
+
+func toPackageDetails(query *osv.Query) (lockfile.PackageDetails, error) {
+	if query.Package.PURL != "" {
+		pkg, err := models.PURLToPackage(query.Package.PURL)
+
+		if err != nil {
+			return lockfile.PackageDetails{}, err
+		}
+
+		return lockfile.PackageDetails{
+			Name:      pkg.Name,
+			Version:   pkg.Version,
+			Ecosystem: lockfile.Ecosystem(pkg.Ecosystem),
+			CompareAs: lockfile.Ecosystem(pkg.Ecosystem),
+		}, nil
+	}
+
+	return lockfile.PackageDetails{
+		Name:      query.Package.Name,
+		Version:   query.Version,
+		Commit:    query.Commit,
+		Ecosystem: query.Package.Ecosystem,
+		CompareAs: query.Package.Ecosystem,
+	}, nil
+}
+
+func setupLocalDBDirectory(localDBPath string) (string, error) {
+	var err error
+	var explicitPath bool
+
+	if localDBPath == "" {
+		if p, envSet := os.LookupEnv(envKeyLocalDBCacheDirectory); envSet {
+			localDBPath = p
+		} else {
+			localDBPath, err = os.UserCacheDir()
+
+			if err != nil {
+				localDBPath = os.TempDir()
+			}
+		}
+	} else {
+		explicitPath = true
+	}
+
+	err = os.Mkdir(path.Join(localDBPath, "osv-scanner"), 0750)
+
+	// if the scanner cannot create its subdirectory when an explicit local db path
+	// has been provided, then it should error rather than fallback to another path
+	//
+	// otherwise, it should fall back to the temp directory before erroring
+	if err != nil && explicitPath {
+		return "", err
+	} else if localDBPath == os.TempDir() {
+		localDBPath = os.TempDir()
+
+		err = os.Mkdir(path.Join(localDBPath, "osv-scanner"), 0750)
+
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return path.Join(localDBPath, "osv-scanner"), nil
+}
+
+func Check(r reporter.Reporter, query osv.BatchedQuery, offline bool, localDBPath string) (*osv.HydratedBatchedResponse, error) {
```

Comment:
nit: `Check` is pretty generic. Is there a more specific name we can give this function?

Reply:
We could call it something like `CheckForVulnerabilities` but that feels pretty verbose vs names elsewhere in the scanner - while I agree `offline.Check` is still more generic than `osvscanner.DoScan`, I don't think it's as big of a problem given its a private package.

Maybe for symmetry we could rename it `MakeRequest` to match the `osv` package, so long as you don't think it'll be confusing do to `offline.MakeRequest`.

Reply:
`MakeRequest` does still seems better here to me.